engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+		return errwrap.Wrapf(err, "invalid serverconfig")
+	}
+	return nil
+}
+
+// getRescueKeys builds a list of keys to be enabled for rescue mode over SSH.
+// ServerRescueSSHKeys provides the selected keys as []string by name. The
+// corresponding data is collected with the Hetzner client (if valid). The
+// resulting keylist is formatted as []*hcloud.SSHKey for use with hcloud, and
+// saved for later use in private field rescueKeys.
+// TODO: standardize this so that it can also be used for serverconfig keys.
+func (obj *HetznerVMRes) getRescueKeys(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getRescueKeys()")
+	}
+	var keylist []*hcloud.SSHKey
+	for _, keyname := range obj.ServerRescueSSHKeys {
+		key, _, err := obj.client.SSHKey.GetByName(ctx, keyname)
+		if err != nil {
+			return errwrap.Wrapf(err, "SSHKey GetByName(%s) failed", keyname)
+		}
+		if key == nil {
+			return fmt.Errorf("unknown keyname: %s", keyname)
+		}
+		if obj.init.Debug {
+			obj.init.Logf("append known keyname: %s", key)
+		}
+		keylist = append(keylist, key)
+	}
+	obj.rescueKeys = keylist
+	return nil
+}
+
+// waitFor provides a general function that returns nil when the provided exit
+// function is satisfied. It retries every WaitInterval until the condition is
+// satisfied. It can exit early in case the WaitTimeout is reached, the context
+// is cancelled or an error occurs. The exit function must be of type
+// HetznerVMCondition, and check a well-defined condition for the resource.
+func (obj *HetznerVMRes) waitFor(ctx context.Context, exitFunc HetznerVMCondition) error {
+	if obj.init.Debug {
+		obj.init.Logf("waitFor()")
+	}
+	var err error = nil
+	conditionSatisfied := false
+	timeout := time.After(time.Duration(obj.WaitTimeout) * time.Second)
+	for {
+		// Get up-to-date serverinfo.
+		if err := obj.getServerUpdate(ctx); err != nil {
+			return errwrap.Wrapf(err, "failed serverupdate request")
+		}
+		// Check if the provided exit condition is satisfied.
+		conditionSatisfied, err = exitFunc(obj)
+		if err != nil {
+			return errwrap.Wrapf(err, "failed to confirm exit condition")
+		}
+		if conditionSatisfied {
+			return nil
+		}
+		// Retry every WaitInterval until the exit condition is satisfied.
+		// Can exit early by timeout, context cancellation or an error.
+		select {
+		case <-time.After(time.Duration(obj.WaitInterval) * time.Second):
+			// retry confirmation
+		case <-timeout:
+			return fmt.Errorf("timeout: exit condition not confirmed after %d seconds", obj.WaitTimeout)
+		case <-ctx.Done():
+			return fmt.Errorf("wait interrupted by context")
+		}
+	}
+}
+
+// HetznerVMCondition provides a HetznerVM-specific type to define functions
+// that check a logical condition on a given HetznerVMRes. Return true if the
+// condition is satisfied, false otherwise. An error must be returned if the
+// condition's input requirements (if any) are not met.
+type HetznerVMCondition func(*HetznerVMRes) (bool, error)
```

Comment:
Not sure this needs a type. You can nuke this type entirely unless there's a reason not to. It only exists in one place...

Reply:
I've been using this type to define an input type for waitFor, to avoid misuse. Can this be done as safely without defining a type? I was having some trouble with these kinds of golang-specific constructs, please let me know if I'm missing something. Will keep the type for now, but will look for a better solution when I make the requested changes to waitFor.